core/src/main/scala/kafka/log/Log.scala

```diff
@@ -473,29 +473,72 @@ class Log(@volatile var dir: File,
         val segment = maybeRoll(messagesSize = validRecords.sizeInBytes,
           maxTimestampInMessages = appendInfo.maxTimestamp,
           maxOffsetInMessages = appendInfo.lastOffset)
+        var isDuplicate = false
+        for ((pid, entry) <- appendInfo.pidEntryMap) {
+          try {
+            // verify that the epoch and sequence are correct before appending. Note that for requests coming from
+            // producer, there will be exactly one LogEntry per Record, and hence one Pid per Record.
+            trace(s"checking sequence for pid: ${pid}, seq: ${entry._1.seq}, end seq: ${entry._2.seq}")
+            pidMap.checkSeqAndEpoch(pid, entry._1.seq, entry._1.epoch)
+          } catch {
+            case e: DuplicateSequenceNumberException =>
+              isDuplicate = true
+              pidMap.entryForPid(pid) match {
+                case Some(pidEntry) =>
+                  if (entry._1.seq < pidEntry.seq) {
+                    // the duplicate is not at the tail of the log. This should never happen, with a properly written client.
+                    // It should also never happen in the replication case.
+                    // Since this is a error situation throw a fatal exception.
+                    throw new IllegalStateException("Detected a duplicate entry in the middle of the log. This suggests a misconfigured or buggy client. Please see __insert_url__ to understand how clients must be configured when running withe enable.idempotence = true");
```

Comment:
This would go back to the user as an UNKNOWN error. Maybe we should use `OutOfOrderSequenceException`?